File: packages/runtime-web-three/src/pathSampling.ts
export type Vec3Tuple = readonly [number, number, number];
export type EasingKind = "linear" | "easeInQuad" | "easeOutQuad" | "easeInOutQuad";

export function ease(kind: EasingKind, t: number): number {
  const clamped = Math.min(1, Math.max(0, t));
  if (kind === "easeInQuad") {
    return clamped * clamped;
  }
  if (kind === "easeOutQuad") {
    return clamped * (2 - clamped);
  }
  if (kind === "easeInOutQuad") {
    return clamped < 0.5 ? 2 * clamped * clamped : 1 - (-2 * clamped + 2) ** 2 / 2;
  }
  return clamped;
}

export function sampleLine(from: Vec3Tuple, to: Vec3Tuple, steps: number, easing: EasingKind = "linear"): Vec3Tuple[] {
  return sampleSteps(steps, (t) => lerpVec3(from, to, ease(easing, t)));
}

export function sampleQuadraticBezier(from: Vec3Tuple, control: Vec3Tuple, to: Vec3Tuple, steps: number, easing: EasingKind = "linear"): Vec3Tuple[] {
  return sampleSteps(steps, (t) => {
    const u = ease(easing, t);
    const a = lerpVec3(from, control, u);
    const b = lerpVec3(control, to, u);
    return lerpVec3(a, b, u);
  });
}

export function sampleCubicBezier(from: Vec3Tuple, controlA: Vec3Tuple, controlB: Vec3Tuple, to: Vec3Tuple, steps: number, easing: EasingKind = "linear"): Vec3Tuple[] {
  return sampleSteps(steps, (t) => {
    const u = ease(easing, t);
    const a = lerpVec3(from, controlA, u);
    const b = lerpVec3(controlA, controlB, u);
    const c = lerpVec3(controlB, to, u);
    return lerpVec3(lerpVec3(a, b, u), lerpVec3(b, c, u), u);
  });
}

export function sampleCatmullRom(points: readonly Vec3Tuple[], stepsPerSegment: number): Vec3Tuple[] {
  if (points.length < 2) {
    return [...points];
  }
  const samples: Vec3Tuple[] = [];
  for (let index = 0; index < points.length - 1; index += 1) {
    const p0 = points[Math.max(0, index - 1)] as Vec3Tuple;
    const p1 = points[index] as Vec3Tuple;
    const p2 = points[index + 1] as Vec3Tuple;
    const p3 = points[Math.min(points.length - 1, index + 2)] as Vec3Tuple;
    const segment = sampleSteps(stepsPerSegment, (t) => catmullRomPoint(p0, p1, p2, p3, t));
    samples.push(...(index === 0 ? segment : segment.slice(1)));
  }
  return samples;
}

function sampleSteps(steps: number, sampler: (t: number) => Vec3Tuple): Vec3Tuple[] {
  const count = Math.max(1, Math.floor(steps));
  return Array.from({ length: count + 1 }, (_, index) => sampler(index / count));
}

function lerpVec3(from: Vec3Tuple, to: Vec3Tuple, t: number): Vec3Tuple {
  return [
    from[0] + (to[0] - from[0]) * t,
    from[1] + (to[1] - from[1]) * t,
    from[2] + (to[2] - from[2]) * t,
  ];
}

function catmullRomPoint(p0: Vec3Tuple, p1: Vec3Tuple, p2: Vec3Tuple, p3: Vec3Tuple, t: number): Vec3Tuple {
  const t2 = t * t;
  const t3 = t2 * t;
  return [0, 1, 2].map((axis) => {
    const v0 = p0[axis] as number;
    const v1 = p1[axis] as number;
    const v2 = p2[axis] as number;
    const v3 = p3[axis] as number;
    return 0.5 * ((2 * v1) + (-v0 + v2) * t + (2 * v0 - 5 * v1 + 4 * v2 - v3) * t2 + (-v0 + 3 * v1 - 3 * v2 + v3) * t3);
  }) as [number, number, number];
}

File: runtime-bevy/crates/threenative_runtime/src/path_sampling.rs
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum EasingKind {
    Linear,
    EaseInQuad,
    EaseOutQuad,
    EaseInOutQuad,
}

pub fn ease(kind: EasingKind, t: f32) -> f32 {
    let clamped = t.clamp(0.0, 1.0);
    match kind {
        EasingKind::EaseInQuad => clamped * clamped,
        EasingKind::EaseOutQuad => clamped * (2.0 - clamped),
        EasingKind::EaseInOutQuad => {
            if clamped < 0.5 {
                2.0 * clamped * clamped
            } else {
                1.0 - (-2.0 * clamped + 2.0).powi(2) / 2.0
            }
        }
        EasingKind::Linear => clamped,
    }
}

pub fn sample_line(
    from: [f32; 3],
    to: [f32; 3],
    steps: usize,
    easing: EasingKind,
) -> Vec<[f32; 3]> {
    sample_steps(steps, |t| lerp_vec3(from, to, ease(easing, t)))
}

pub fn sample_quadratic_bezier(
    from: [f32; 3],
    control: [f32; 3],
    to: [f32; 3],
    steps: usize,
    easing: EasingKind,
) -> Vec<[f32; 3]> {
    sample_steps(steps, |t| {
        let u = ease(easing, t);
        let a = lerp_vec3(from, control, u);
        let b = lerp_vec3(control, to, u);
        lerp_vec3(a, b, u)
    })
}

pub fn sample_cubic_bezier(
    from: [f32; 3],
    control_a: [f32; 3],
    control_b: [f32; 3],
    to: [f32; 3],
    steps: usize,
    easing: EasingKind,
) -> Vec<[f32; 3]> {
    sample_steps(steps, |t| {
        let u = ease(easing, t);
        let a = lerp_vec3(from, control_a, u);
        let b = lerp_vec3(control_a, control_b, u);
        let c = lerp_vec3(control_b, to, u);
        lerp_vec3(lerp_vec3(a, b, u), lerp_vec3(b, c, u), u)
    })
}

pub fn sample_catmull_rom(points: &[[f32; 3]], steps_per_segment: usize) -> Vec<[f32; 3]> {
    if points.len() < 2 {
        return points.to_vec();
    }
    let mut samples = Vec::new();
    for index in 0..points.len() - 1 {
        let p0 = points[index.saturating_sub(1)];
        let p1 = points[index];
        let p2 = points[index + 1];
        let p3 = points[(index + 2).min(points.len() - 1)];
        let segment = sample_steps(steps_per_segment, |t| catmull_rom_point(p0, p1, p2, p3, t));
        if index == 0 {
            samples.extend(segment);
        } else {
            samples.extend(segment.into_iter().skip(1));
        }
    }
    samples
}

fn sample_steps(steps: usize, sampler: impl Fn(f32) -> [f32; 3]) -> Vec<[f32; 3]> {
    let count = steps.max(1);
    (0..=count)
        .map(|index| sampler(index as f32 / count as f32))
        .collect()
}

fn lerp_vec3(from: [f32; 3], to: [f32; 3], t: f32) -> [f32; 3] {
    [
        from[0] + (to[0] - from[0]) * t,
        from[1] + (to[1] - from[1]) * t,
        from[2] + (to[2] - from[2]) * t,
    ]
}

fn catmull_rom_point(p0: [f32; 3], p1: [f32; 3], p2: [f32; 3], p3: [f32; 3], t: f32) -> [f32; 3] {
    let t2 = t * t;
    let t3 = t2 * t;
    let mut output = [0.0; 3];
    for axis in 0..3 {
        let v0 = p0[axis];
        let v1 = p1[axis];
        let v2 = p2[axis];
        let v3 = p3[axis];
        output[axis] = 0.5
            * ((2.0 * v1)
                + (-v0 + v2) * t
                + (2.0 * v0 - 5.0 * v1 + 4.0 * v2 - v3) * t2
                + (-v0 + 3.0 * v1 - 3.0 * v2 + v3) * t3);
    }
    output
}
